### PR TITLE
Move flattened field to core.

### PR DESCRIPTION
**Pull request by jtibshirani**
This field mapper only lived in its own module so it could be licensed as x-pack
basic. Now it can be moved to core, which matches its status as a core type.

This commit doesn't bring major code simplification yet, but it does improve
how the tests are factored.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/IndicesModule.java [hunk 1]
```diff
@@ -107,18 +108,21 @@ public IndicesModule(List<MapperPlugin> mapperPlugins) {
...
+
         DateFieldMapper.Resolution milliseconds = DateFieldMapper.Resolution.MILLISECONDS;
         mappers.put(milliseconds.type(), DateFieldMapper.MILLIS_PARSER);
         DateFieldMapper.Resolution nanoseconds = DateFieldMapper.Resolution.NANOSECONDS;
         mappers.put(nanoseconds.type(), DateFieldMapper.NANOS_PARSER);
-        mappers.put(IpFieldMapper.CONTENT_TYPE, IpFieldMapper.PARSER);
-        mappers.put(TextFieldMapper.CONTENT_TYPE, TextFieldMapper.PARSER);
+
+        mappers.put(FieldAliasMapper.CONTENT_TYPE, new FieldAliasMapper.TypeParser());
+        mappers.put(FlattenedFieldMapper.CONTENT_TYPE, FlattenedFieldMapper.PARSER);
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/indices/IndicesModule.java:119]: Here I added `FlattenedFieldMapper` and tried to alphabetize the entries.

**Comment by jtibshirani** [comment]
A couple notes:
* I introduced a new package `org.elasticsearch.index.mapper.flattened`. This is not common practice yet, but felt appropriate for more complex like `flattened` which have various helper classes.
* When moving `flattened` to its own module, we introduced an unnecessary abstraction `DynamicKeyFieldMapper`. I had planned to remove this while moving `flattened` back to core, but ended up keeping it. I now find the abstraction pretty clean, and it may be helpful for other field types we're considering like numeric flattened fields.

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/340_flattened.yml [hunk 2]
```diff
@@ -1,9 +1,10 @@
----
-"Test exists query on flattened object field":
+setup:
   - skip:
-      version: " - 7.2.99"
-      reason: "Flattened fields were implemented in 7.3."
+      version: " - 7.99.99"
```
Comment (mayya-sharipova) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search/340_flattened.yml:3]: Version change because some nodes may now have x-pack installed?

**Review by mayya-sharipova — APPROVED** [review]
Thanks @jtibshirani, makes sense. +1 for a new package `mapper.flattened`.

One extra question I have, do we want to remove `x-pack` reference in  `docs/reference/mapping/types/flattened.asciidoc` as well?

File: rest-api-spec/src/main/resources/rest-api-spec/test/search/340_flattened.yml (hunk 2 above)
Comment (jtibshirani) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search/340_flattened.yml:3]: These core REST tests generally do not run with x-pack installed. So I bumped the version here, and plan to adjust to cover 7.12 once it's backported.

**Comment by jtibshirani** [comment]
> do we want to remove x-pack reference in docs/reference/mapping/types/flattened.asciidoc as well?

I think it makes sense to remove this, I pushed a new commit.

Thanks for the review. Before merging I'd like to get more clarity on our overall design (for example when should a field type live in core vs. mapper-extras?) I'll loop back here with conclusions.

**Pull request merged**